components/brave_wallet/browser/solana_data_decoder_utils.cc

```diff
@@ -0,0 +1,204 @@
+  return Base58Encode(std::vector<uint8_t>(
+      input.begin() + offset - kSolanaPubkeySize, input.begin() + offset));
+}
+
+absl::optional<std::string> DecodeOptionalPublicKey(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  if (offset == input.size()) {
+    return absl::nullopt;
+  }
+
+  // First byte is 0 or 1 to indicate if public key is passed.
+  // And the rest bytes are the actual public key.
+  if (input[offset] == 0) {
+    offset++;
+    return "";  // No public key is passed.
+  } else if (input[offset] == 1) {
+    offset++;
+    return DecodePublicKey(input, offset);
+  } else {
+    return absl::nullopt;
+  }
+}
+
+// bincode::serialize uses two u32 together for the string length and a byte
+// array for the actual strings. The first u32 represents the lower bytes of
+// the length, the second represents the upper bytes. The upper bytes will have
+// non-zero value only when the length exceeds the maximum of u32.
+// We currently cap the length here to be the max size of std::string
+// on 32 bit systems, it's safe to do so because currently we don't expect any
+// valid cases would have strings larger than it.
+absl::optional<std::string> DecodeString(const std::vector<uint8_t>& input,
+                                         size_t& offset) {
+  auto len_lower = DecodeUint32(input, offset);
+  if (!len_lower || *len_lower > kMaxStringSize32Bit)
+    return absl::nullopt;
+  auto len_upper = DecodeUint32(input, offset);
+  if (!len_upper || *len_upper != 0) {  // Non-zero means len exceeds u32 max.
+    return absl::nullopt;
+  }
+
+  if (offset + *len_lower > input.size())
+    return absl::nullopt;
+
+  offset += *len_lower;
+  return std::string(reinterpret_cast<const char*>(&input[offset - *len_lower]),
+                     *len_lower);
+}
+
+// Expects a the bytes of a Borsh encoded Metadata struct (see
+// https://docs.rs/spl-token-metadata/latest/spl_token_metadata/state/struct.Metadata.html)
+// and returns the URI string in of the nested Data struct (see
+// https://docs.rs/spl-token-metadata/latest/spl_token_metadata/state/struct.Data.html)
+// as a GURL.
+absl::optional<GURL> DecodeMetadataUri(const std::vector<uint8_t> data) {
+  size_t offset = 0;
+  offset = offset + /* Skip first byte for metadata.key */ 1 +
+           /* Skip next 32 bytes for `metadata.update_authority` */ 32 +
+           /* Skip next 32 bytes for `metadata.mint` */ 32;
+
+  // Skip next field, metdata.data.name, a string
+  // whose length is represented by a leading 32 bit integer
+  auto length = DecodeUint32(data, offset);
+  if (!length) {
+    return absl::nullopt;
+  }
+  offset += static_cast<size_t>(*length);
+
+  // Skip next field, `metdata.data.symbol`, a string
+  // whose length is represented by a leading 32 bit integer
+  length = DecodeUint32(data, offset);
+  if (!length) {
+    return absl::nullopt;
+  }
+  offset += static_cast<size_t>(*length);
+
+  // Parse next field, metadata.data.uri, a string
```

Comment:
Actually you should just call DecodeString for this, which does the right thing and check size too.

Reply:
Now I'm confused why your case seems different than Solana transaction data.

Reply:
Ok they're probably different cases in the first place, different serialize method it seems.
As mentioned in the DM, plz leave instruction decoder intact, and use your own DecodeMetadataUri in your upcoming NftMetadataFetcher class or the place you're going to use this function.
Make sure size check is performed whenever you're going to access the arr.